### ISSUE TITLE
Allow reflected functions to use non-reflected functions

### ISSUE THREAD

**Issue by facundominguez** [ucsd-progsys/liquidhaskell#1892]
In stitch I was trying at some point to reflect the stitch type checker. However, the type checker uses `Data.Map.lookup` to find globally defined symbols.

Now, Liquid Haskell would insist that `Data.Map.lookup` needs to be reflected in order to reflect the type checker, when all it actually needs is a function in the logic to stand for it. I'm not interested in reflecting `Data.Map.lookup` since it is rather non-trivial and I'm happy to introduce assumptions about its behavior as necessary.

My workaround [here](https://github.com/facundominguez/stitch-lh/blob/c411889d2816954e69e8fda1671711b9ed2600cc/src/Language/Stitch/LH/Data/Map.hs#L32) implements a lookup function that doesn't terminate, but nonetheless can be reflected. And then I use rewrite rules to convince GHC to replace the non-terminating function with the actual `lookup` function from `Data.Map`.

This is a horrible hack as far as hacks go. Could we instead just tell LH to put in the logic an uninterpreted function
```
{-@ opaque-reflect Data.Map.lookup :: k -> Map k v -> Maybe v @-}
```
?
Now when LH finds that the stitch type-checker uses `Data.Map.lookup`, it uses the uninterpreted function without further ado.

One other use case for a feature like this is in the length function of lists and bytestrings, and many other functions. `liquid-base` defines a measure
```
instance measure len :: forall a. [a] -> GHC.Types.Int
  len []     = 0
  len (y:ys) = 1 + len ys
```
Then length is defined as
```
{-@ length :: xs:[a] -> {v: GHC.Types.Int | v = len(xs)} @-}
```
This still doesn't allow to reflect functions that call to `length` despite of being a perfectly usable `len` function in the logic. Moreover, it requires the user to translate between the two names when crossing from code to specification.

Why not just name the measure the same as the function? And then let LH translate between the two when reflecting.

**Comment by ranjitjhala** [comment]
Yes this is an *excellent* idea - in fact opaque-measures can be useful just to write specs.

**Comment by facundominguez** [comment]
It struck me last weekend that perhaps opaque reflect should rather be the default. If a function is not available in the logic when reflecting some other function, have LH reflect it opaquely.

**Comment by nikivazou** [comment]
I agree, and actually do that a lot, manually. 
If I want to reflect `foo` that calls `bar` and cannot reflect `bar`, then the hack around it is: 

```
foo = .... bar 

{-@ measure  bar :: a -> b @-}
{-@ assume bar :: x:a -> {v:b |  v = bar x } @-}
bar = unreflected 
```

**Comment by ranjitjhala** [comment]
The benefit of @facundo’s proposal is we could do this also for imported
functions (eg in the prelude) without having to redefine them…
